pkg/storage/stores/tsdb/store.go

```diff
@@ -17,37 +19,69 @@ import (
 	util_log "github.com/grafana/loki/pkg/util/log"
 )
 
-// we do not need to build store for each schema config since we do not do any schema specific handling yet.
-// If we do need to do schema specific handling, it would be a good idea to abstract away the handling since
-// running multiple head managers would be complicated and wasteful.
-var storeInstance *store
-
 type store struct {
-	indexWriter IndexWriter
-	indexStore  series.IndexStore
+	indexShipper indexshipper.IndexShipper
+	indexWriter  IndexWriter
+	indexStore   series.IndexStore
+	stopOnce     sync.Once
 }
 
+type newStoreFactoryFunc func(
+	indexShipperCfg indexshipper.Config,
+	p config.PeriodConfig,
+	f *fetcher.Fetcher,
+	objectClient client.ObjectClient,
+	limits downloads.Limits,
+	tableRanges config.TableRanges,
+	reg prometheus.Registerer,
+) (
+	chunkWriter stores.ChunkWriter,
+	indexStore series.IndexStore,
+	stopFunc func(),
+	err error,
+)
+
 // NewStore creates a new store if not initialized already.
 // Each call to NewStore will always build a new stores.ChunkWriter even if the store was already initialized since
 // fetcher.Fetcher instances could be different due to periodic configs having different types of object storage configured
 // for storing chunks.
-func NewStore(indexShipperCfg indexshipper.Config, p config.PeriodConfig, f *fetcher.Fetcher,
-	objectClient client.ObjectClient, limits downloads.Limits, tableRanges config.TableRanges, reg prometheus.Registerer) (stores.ChunkWriter, series.IndexStore, error) {
-	if storeInstance == nil {
-		storeInstance = &store{}
-		err := storeInstance.init(indexShipperCfg, objectClient, limits, tableRanges, reg)
-		if err != nil {
-			return nil, nil, err
+// It also helps us make tsdb store a singleton because
+// we do not need to build store for each schema config since we do not do any schema specific handling yet.
+// If we do need to do schema specific handling, it would be a good idea to abstract away the handling since
+// running multiple head managers would be complicated and wasteful.
+var NewStore = func() newStoreFactoryFunc {
+	var storeInstance *store
+	return func(
+		indexShipperCfg indexshipper.Config,
+		p config.PeriodConfig,
+		f *fetcher.Fetcher,
+		objectClient client.ObjectClient,
+		limits downloads.Limits,
+		tableRanges config.TableRanges,
+		reg prometheus.Registerer,
+	) (
+		stores.ChunkWriter,
+		series.IndexStore,
+		func(),
+		error,
+	) {
+		if storeInstance == nil {
+			storeInstance = &store{}
+			err := storeInstance.init(indexShipperCfg, objectClient, limits, tableRanges, reg)
+			if err != nil {
+				return nil, nil, nil, err
+			}
 		}
-	}
 
-	return NewChunkWriter(f, p, storeInstance.indexWriter), storeInstance.indexStore, nil
-}
+		return NewChunkWriter(f, p, storeInstance.indexWriter), storeInstance.indexStore, storeInstance.Stop, nil
```

Comment:
Why not return the `storeInstance`?

Reply:
For which type? We can't do it for writer atleast as explained here https://github.com/grafana/loki/blob/d6f0455b656d9ad2c99aabcda5ec2dbbbd6135a8/pkg/storage/stores/tsdb/store.go#L45

Reply:
@jeschkies Yeah, we should refactor to make this more intuitive/less ugly.